### ISSUE TITLE
[TagHelper] React To #298, TagHelpers does not work when TagHelperOutput.SelfClose == true

### ISSUE THREAD

**Issue by ianhong** [aspnet/Razor#300]
Currently TagHelperOutput.SelfClose is always false by default, but when #298 is checked in, TagHelperOutput.SelfClose will be set to true if the original tag input is self-closing such as `<div asp-validation-summary="ValidationSummary.All" class="order" />`

If the TagHelperOutput.SelfClose is set to true, the tag helpers generating inner elements don't work. 

For example,
`<div asp-validation-summary="ValidationSummary.All" class="order" />`
should generate the following HTMLs:
`<div class="order validation-summary-errors" data-valmsg-summary="true"><ul><li>A value is required.</li>`
`<li>The Password field is required.</li>`
`</ul></div>`

But because TagHelperOutput returns string.empty when the selfclosing is true as follows:, it doesn't work.
https://github.com/aspnet/Razor/blob/dev/src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs#L161-L164

For example, it will simply return
`<div class="order validation-summary-errors" data-valmsg-summary="true"/>`

**Comment by ianhong** [comment]
@NTaylorMullen I opened this issue that we discussed this morning.

**Comment by dougbu** [comment]
@ianhong this sounds more like a "react to #298" task that should be open in MVC.  in particular what won't work in the described scenario once the validation summary tag helper sets `SelfClosing` to `false`?

**Comment by ianhong** [comment]
@dougbu 
Thanks, yes I should change the title, Yes you are correct, this bug can be fixed from the individual taghelpers too. but I think we should be able to do something more here since
1) My understanding is not only MVC TagHelpers but also custom TagHelpers have similar problems. 
2) I think TagHelperOutput could be a little bit more intelligent so that it could set SelfClosing=true when it sees an nested tags from the content.

**Comment by ianhong** [comment]
Adding @yishaigalatzer

**Comment by yishaigalatzer** [comment]
@ianhong not sure why this is a separate issue from 298. 298 is just in the design stage, there is one fix for this, not two unrelated ones

**Comment by ianhong** [comment]
@yishaigalatzer Agreed. @NTaylorMullen what's your opinion about combining them for now? If needed, I can open an react issue for MVC tag helpers later.

**Comment by dougbu** [comment]
@ianhong you need to open a react __PR__ for MVC and that must be pushed at the same time as the final #298 commits.  no need for a separate issue unless you feel strongly about
> `TagHelperOutput` could be a little bit more intelligent

("intelligence" is often too magical. perhaps the `[Pre|Post]?Content` properties could change `SelfClosing` to `false`.  but second-guessing the tag helper author when the `GenerateFoo()` methods are called would be a Bad Thing:tm:  in any case, this is a design discussion -- an idea you should either include in #298 or bin.)

**Comment by ianhong** [comment]
@dougbu got it. By the way, another interesting case will be when TagName is set to null or empty. I saw a test case doing it, but I'm not sure of the actual scenario of this. @NTaylorMullen could you give us an example of this??

**Comment by NTaylorMullen** [comment]
@ianhong sorry to have mislead you into creating this separate issue. @dougbu @yishaigalatzer and I talked and the fix to this issue will be included in your #298 fix.

There will be two parts:

1. Razor fixes to `TagHelperOutput` - Schedule a design discussion on this, we'll need to discuss a good way to do this.
1. MVC fixes to ensure that all of our `TagHelper`s are being explicit on whether they're self-closing or not.

Closing for now since this will be done in #298
